Gems/ROS2/Code/Source/Sensor/ROS2SensorComponent.cpp

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+#include "Sensor/ROS2SensorComponent.h"
+#include "ROS2/ROS2Bus.h"
+#include "Utilities/ROS2Names.h"
+
+#include <AzCore/Component/Entity.h>
+#include <AzCore/Serialization/SerializeContext.h>
+
+namespace ROS2
+{
+    void ROS2SensorComponent::Activate()
+    {
+        AZ::TickBus::Handler::BusConnect();
+    }
+
+    void ROS2SensorComponent::Deactivate()
+    {
+        AZ::TickBus::Handler::BusDisconnect();
+    }
+
+    void ROS2SensorComponent::Reflect(AZ::ReflectContext* context)
+    {
+        if (AZ::SerializeContext* serialize = azrtti_cast<AZ::SerializeContext*>(context))
+        {
+            serialize->Class<ROS2SensorComponent, AZ::Component>()
+                ->Version(1)
+                ->Field("SensorConfiguration", &ROS2SensorComponent::m_sensorConfiguration)
+                ;
+        }
+    }
+
+    const SensorConfiguration& ROS2SensorComponent::GetConfiguration() const
+    {
+        return m_sensorConfiguration;
+    }
+
+    AZStd::string ROS2SensorComponent::GetNamespace() const
+    {
+        // TODO - hold frame?
+        auto ros2Frame = GetEntity()->FindComponent<ROS2FrameComponent>();
+        return ros2Frame->GetNamespace();
+    };
+
+    AZStd::string ROS2SensorComponent::GetFullTopic() const
+    {
+        return ROS2Names::GetNamespacedName(GetNamespace(), m_sensorConfiguration.m_topic);
+    }
+
+    AZStd::string ROS2SensorComponent::GetFrameID() const
+    {
+        auto ros2Frame = GetEntity()->FindComponent<ROS2FrameComponent>();
+        return ros2Frame->GetFrameID();
+    }
+
+    void ROS2SensorComponent::GetRequiredServices(AZ::ComponentDescriptor::DependencyArrayType& required)
+    {
+        required.push_back(AZ_CRC("ROS2Frame"));
```

Comment:
My thought, which I think will be our bone of contention :smile: 

I would lose this requirement. As for the current `ROS2SensorComponent` implementation, it is just a proxy for the `ROS2FrameComponent` ("get" method names are the same).

I think it would be better to give Sensors the decision whether they require frame components or not and let them use the `ROS2FrameComponent` directly.

As for the `namespace` ownership problem: in my opinion, the concept of a namespace is close both to `tf` and `topic` (and even `node`, for example), therefore, I would move it somewhere else, outside both frame and sensor components.

Then `ROS2SensorComponent::GetNamespace()` would look for the "closest" namespace owner (namespaceComponent?) and return some namespace or empty string when there is nothing found.

Reply:
I disagree: I believe that each sensor should have a frame. I can't picture a sensor which should not have a frame.
ROS2SensorComponent is not just a proxy for ROS2FrameComponent. It represents different concepts and has distinct members (for example, ros2 publisher concept, which will be encapsulated in the future but now is topic, frequency, option to publish. Also, whether to visualise data. Also, controlling execution (FrequencyTick())). 

I also disagree that namespace ownership should be somewhere else - but we can discuss that when namespaces PR is cast.